examples/workshop-fomo-generalized/index.rsh

```diff
@@ -0,0 +1,68 @@
+'reach 0.1';
+
+// FOMO Workshop generalized to last N winners
+const NUM_OF_WINNERS = 3;
+
+const CommonInterface = {
+  showOutcome: Fun([Array(Address, NUM_OF_WINNERS)], Null),
+};
+
+const FunderInterface = {
+  ...CommonInterface,
+  getParams: Fun([], Object({
+    deadline: UInt, // relative deadline
+    ticketPrice: UInt,
+  })),
+};
+
+const BuyerInterface = {
+  ...CommonInterface,
+  shouldBuyTicket: Fun([UInt], Bool),
+  showPurchase: Fun([Address], Null),
+};
+
+export const main = Reach.App(
+  { },
+  [
+    ['Funder', FunderInterface], ['class', 'Buyer', BuyerInterface],
+  ],
+  (Funder, Buyer) => {
+    const showOutcome = (winners) =>
+      each([Funder, Buyer], () => interact.showOutcome(winners));
+
+    Funder.only(() => {
+      const { ticketPrice, deadline } = declassify(interact.getParams()); });
+    Funder.publish(ticketPrice, deadline);
+
+    const initialWinners = Array.replicate(NUM_OF_WINNERS, Funder);
+
+    // Until deadline, allow buyers to buy ticket
+    const [ keepGoing, winners, ticketsSold ] =
+      parallel_reduce([ true, initialWinners, 0 ])
+        .invariant(balance() == ticketsSold * ticketPrice)
+        .while(keepGoing)
+        .case(
+          Buyer,
+          (() => ({
+            when: declassify(interact.shouldBuyTicket(ticketPrice)) })),
+          (() => ticketPrice),
+          () => {
+            const buyer = this;
+            Buyer.only(() => interact.showPurchase(buyer));
+            const idx = ticketsSold % NUM_OF_WINNERS;
```

Comment:
You might want to comment in the workshop text about why this implements a ring buffer